deepspeed/runtime/pipe/engine.py

```diff
@@ -889,10 +913,20 @@ def _exec_send_grads(self, buffer_id):
 
         # Partition the gradient
         if self.is_grad_partitioned:
-            assert torch.is_tensor(inputs)
-            part = PartitionedTensor(tensor=inputs.grad,
+            if isinstance(inputs, tuple):
+                first_input = inputs[0]
+                assert all([torch.is_tensor(elt) for elt in inputs[1:]])
+                inputs_grad_tail = [elt.grad for elt in inputs[1:] if elt.grad is not None]
+            elif torch.is_tensor(inputs):
+                first_input = inputs
+                inputs_grad_tail = []
+            else:
+                raise NotImplementedError
+            assert torch.is_tensor(first_input)
+            part = PartitionedTensor(tensor=first_input.grad,
                                      group=self.grid.get_slice_parallel_group())
-            inputs = tuple([part.to_meta(), part.data()])
+
+            inputs = ([part.to_meta(), part.data(), *inputs_grad_tail])
```

Comment:
should this not be a tuple?

Reply:
This is a tuple, in python you don't need to specifcy `tuple`, parenthesis work just fine.

Reply:
ah well played I didn't notice, which now begs the question why it works ...